ctapipe/reco/hillas_intersection.py

```diff
@@ -495,25 +499,27 @@ def reconstruct_xmax(
             np.array(ty),
         )
         weight = np.array(amp)
-        mean_height = np.sum(height * weight) / np.sum(weight)
+        mean_distance = np.sum(height * weight) / np.sum(weight)
 
         # This value is height above telescope in the tilted system,
         # we should convert to height above ground
-        mean_height *= np.cos(zen)
+        mean_height = mean_distance * np.cos(zen.to_value(u.rad))
 
         # Add on the height of the detector above sea level
-        mean_height += 2100  # TODO: replace with instrument info
+        if self.subarray.reference_location is not None:
+            mean_height += self.subarray.reference_location.geodetic.height.to_value(
+                u.m
+            )
+        else:
+            # FIXME: Can remoev this check once we ensure the reference_location is always loaded
+            warnings.warn(
+                "Computing h_max with no reference location. Height will be wrong."
+            )
 
         if mean_height > 100000 or np.isnan(mean_height):
-            mean_height = 100000
-
-        mean_height *= u.m
-        # Lookup this height in the depth tables, the convert Hmax to Xmax
-        # x_max = self.thickness_profile(mean_height.to(u.km))
-        # Convert to slant depth
-        # x_max /= np.cos(zen)
+            mean_height = np.nan
 
-        return mean_height
+        return u.Quantity(mean_height, u.m)
```

Comment:
This is probably a good opportunity to add returning an `h_max_uncert` from this function, even if just giving NaN for now rather than manually setting it when the `ReconstructedGeometryContainer`  is created